### DISCUSSION TITLE
How to implement app menu?

### DISCUSSION THREAD

**Discussion by gtamas** [Relm4/Relm4#550] [General]
How can I create the main menu of the app? I'm talking about the one that's located at the top of the screen in MacOS apps.

I checked the examples, but I'm still not sure.

Should I follow something like this:

https://github.com/Relm4/Relm4/blob/main/examples/popover.rs

This appears to implement a menu button with popover menu, but what I want is a main menu, so I'm not sure if this is the right API to use. I would also like to attach keyboard shortcuts to menu options.

**Comment by AaronErhardt** [comment]
Do you want to implement a menu bar (the traditional menu in a bar at the top of the window) or a header bar according to the [GNOME HIG](https://developer.gnome.org/hig/patterns/containers/header-bars.html)?
Or do you want something else? I'm not familiar with MacOS, maybe a screenshot would help 😅

**Reply by gtamas** [reply]
Yes exactly, I wanna implement the menubar. Under OSX that's located at the top of the screen and looks like this:

<img width="640" alt="image" src="https://github.com/Relm4/Relm4/assets/687850/c90f24c8-3625-41c3-b8a8-944dff179e24">

But yes, this is supposed to be the same menu Windows display at the top of the main app window.

**Reply by AaronErhardt** [reply]
Ok, I haven't done this yet, but if I understand the docs of GTK4 correctly, you need to create a menu model (which stores the information about the menu) and then call [`set_menubar()`](https://gtk-rs.org/gtk4-rs/git/docs/gtk4/prelude/trait.GtkApplicationExt.html#method.set_menubar) on the `gtk::Application` of your app (which you can get with [`main_application()`](https://docs.rs/relm4/latest/relm4/fn.main_application.html)).

You can create the menu model using the `menu!` macro in Relm4 like in this example: https://github.com/Relm4/Relm4/blob/main/examples/menu.rs#L60

**Reply by gtamas** [reply]
OK so I was able to create the model, but it still doesn't work. I have this

```
fn init(
        params: Self::Init,
        root: &Self::Root,
        sender: ComponentSender<Self>,
    ) -> ComponentParts<Self> {  

       let widgets = view_output!();

        relm4::menu! {
            main_menu: {
                custom: "my_widget",
                "Example" => ExampleAction,
                "Example2" => ExampleAction,
                "Example toggle" => ExampleU8Action(1_u8),
                section! {
                    "Section example" => ExampleAction,
                    "Example toggle" => ExampleU8Action(1_u8),
                },
                section! {
                    "Example" => ExampleAction,
                    "Example2" => ExampleAction,
                    "Example Value" => ExampleU8Action(1_u8),
                }
            }
        };

        let app = relm4::main_application();
        
        app.set_accelerators_for_action::<ExampleAction>(&["<primary>W"]);

        let action: RelmAction<ExampleAction> = {
            RelmAction::new_stateless(move |_| {
                println!("Statelesss action!");
      
            })
        };

        let action2: RelmAction<ExampleU8Action> =
            RelmAction::new_stateful_with_target_value(&0, |_, state, _value| {
                *state ^= 1;
                dbg!(state);
            });
        
        widgets.main_window.set_show_menubar(true);
        let mut group = RelmActionGroup::<WindowActionGroup>::new();
        group.add_action(action);
        group.add_action(action2);
        group.register_for_widget(&widgets.main_window);

        app.set_menubar(Some(&main_menu));
}
```

Using the example code listed above, I can add the menu. If I don't call `app.set_menubar(Some(&main_menu));`
then OSX generates a default menu for the app, but once this is called nothing happens. There is no menu, no error nothing. The default one won't be generated either of course, since I just added my own.

Anyway, I'm not sure, does this feature even work with app menus? I mean, most GTK-rs apps I saw just use MenuButtons, but they don't seem to have a real main menu for some reason. Which is weird, by the way, since all apps have such menu under MacOS or Windows, for example.

I checked the GTK-rs docs as well, but they don't mention any missing features or bugs. This should work. Not sure what to try :(

**Reply by gtamas** [reply]
Update: It works, well kinda :)

Turns out that if you create an empty MenuModel manually and then add the model created by ` menu!` as submenu, then it works.

```
 let menu_bar = relm4::gtk::gio::Menu::new();
 menu_bar.append_submenu(Some("Options"), &main_menu);
```

I'm not really sure why this works though. This doesn't seem to be the normal behaviour. Also all the menu options seem to be inactive by default. You can't click them, but the registered keyboard shortcuts work perfectly.

**Reply by AaronErhardt** [reply]
> I mean, most GTK-rs apps I saw just use MenuButtons, but they don't seem to have a real main menu for some reason. Which is weird, by the way, since all apps have such menu under MacOS or Windows, for example.

The GNOME HIG (human interface guidelines) recommend using menu buttons and most apps follow those guidelines.

> Turns out that if you create an empty MenuModel manually and then add the model created by  menu! as submenu, then it works.

You can use the macro for that, too, btw:

```rust
    relm4::menu! {
        main_menu: {
            "Options" {
                "Example" => ExampleAction,
                "Example2" => ExampleAction,
                "Example toggle" => ExampleU8Action(1_u8),
            }
        }
    };
```

> I'm not really sure why this works though. This doesn't seem to be the normal behaviour. 

Now that I think about it, I think it makes sense. Each entry in the menu bar is a submenu, so if you don't add things to a submenu, GTK doesn't know what you're trying to do.

> Also all the menu options seem to be inactive by default. You can't click them, but the registered keyboard shortcuts work perfectly.

You should be able to click the menu entries once you registered an action for them. Otherwise GTK has nothing connected to them and shows them as inactive. But it should work with your code, it does at least on my system (both shortcut and clicking).

**Reply by gtamas** [reply]
OK thanks for the info!

Yes it works fine with the macro now. But the options in the menu are still inactive. I don't get it. Isn't this part is responsible for registering the actions?

```
  let mut group = RelmActionGroup::<WindowActionGroup>::new();
  group.add_action(action);
  group.add_action(action2);
  group.register_for_widget(&widgets.main_window);
```

Do I need to call something else?

I also noticed that actually I can't click to the menu unless I click something else, outside the app first. After that it works. So something else is wrong here..

**Reply by AaronErhardt** [reply]
This code works for me (modified menu example):

```rust
use adw::prelude::{ApplicationWindowExt, GtkApplicationExt};
use gtk::prelude::{BoxExt, ButtonExt, GtkWindowExt, OrientableExt};
use relm4::actions::{AccelsPlus, ActionablePlus, RelmAction, RelmActionGroup};
use relm4::{
    gtk, ComponentParts, ComponentSender, RelmApp, RelmWidgetExt, SimpleComponent,
};

#[derive(Default)]
struct App {
    counter: u8,
}

#[derive(Debug)]
enum Msg {
    Increment,
    Decrement,
}

#[relm4::component]
impl SimpleComponent for App {
    type Init = u8;
    type Input = Msg;
    type Output = ();

    view! {
        #[root]
        main_window = gtk::ApplicationWindow {
            set_title: Some("Menu example"),
            set_default_size: (300, 100),

            gtk::Box {
                set_orientation: gtk::Orientation::Vertical,
                set_margin_all: 5,
                set_spacing: 5,

                gtk::Button {
                    set_label: "Increment",
                    connect_clicked => Msg::Increment,
                    ActionablePlus::set_action::<ExampleU8Action>: 1,
                },
                gtk::Button {
                    set_label: "Decrement",
                    connect_clicked => Msg::Decrement,
                },
                gtk::Label {
                    set_margin_all: 5,
                    #[watch]
                    set_label: &format!("Counter: {}", model.counter),
                },
            },
        },
    }

    fn init(
        _params: Self::Init,
        root: &Self::Root,
        sender: ComponentSender<Self>,
    ) -> ComponentParts<Self> {
        let model = App { counter: 0 };
        let widgets = view_output!();

    relm4::menu! {
        main_menu: {
            "Options" {
                "Example" => ExampleAction,
                "Example2" => ExampleAction,
                "Example toggle" => ExampleU8Action(1_u8),
            }
        }
    };

        let app = relm4::main_application();

        app.set_accelerators_for_action::<ExampleAction>(&["<primary>W"]);

        let action: RelmAction<ExampleAction> = {
            RelmAction::new_stateless(move |_| {
                println!("Statelesss action!");
            })
        };

        let action2: RelmAction<ExampleU8Action> =
            RelmAction::new_stateful_with_target_value(&0, |_, state, _value| {
                *state ^= 1;
                dbg!(state);
            });

        widgets.main_window.set_show_menubar(true);
        let mut group = RelmActionGroup::<WindowActionGroup>::new();
        group.add_action(action);
        group.add_action(action2);
        group.register_for_widget(&widgets.main_window);



        app.set_menubar(Some(&main_menu));

        ComponentParts { model, widgets }
    }

    fn update(&mut self, msg: Self::Input, _sender: ComponentSender<Self>) {
        match msg {
            Msg::Increment => {
                self.counter = self.counter.wrapping_add(1);
            }
            Msg::Decrement => {
                self.counter = self.counter.wrapping_sub(1);
            }
        }
    }
}

relm4::new_action_group!(WindowActionGroup, "win");

relm4::new_stateless_action!(ExampleAction, WindowActionGroup, "example");
relm4::new_stateful_action!(ExampleU8Action, WindowActionGroup, "example2", u8, u8);

fn main() {

    let app = RelmApp::new("relm4.example.menu");
    app.run::<App>(0);
}
```

**Reply by gtamas** [reply]
I can confirm that this does not work on Mac. Compiled the code above and the options are inactive.
This is MacOS 12.3 (Monterey). But I don't think it's a relm4 issue, probably it's GTK related. Or maybe we need to something more to make this work under OSX, the docs don't mention anything special though.

Will post here if I find a solution.

**Reply by AaronErhardt** [reply]
Maybe this is related to this: https://gitlab.gnome.org/GNOME/gtk/-/issues/5010

**Reply by gtamas** [reply]
Thanks I checked this and I think the issue might be related to Relm4 too, after all.

I tried what they described in the GTK issue and it worked. If you place the application binary in the proper location, the initial unresponsiveness of the menu gets resolved. It works fine.

However, the menu options are still inactive, even if I move the binary as they described. BUT: In the GTK4 demo app they mentioned, the menu options work fine! The options are active and clickable either with or without moving the binary, while compiling the rust app (even the menu example code above) results in a menu with inactive items.

Could this be related to the fact that we initialise and register the menu in the  fn init()? From the GTK docs:

https://docs.gtk.org/gtk4/method.Application.set_menubar.html

"This can only be done in the primary instance of the application, after it has been registered. GApplication::startup is a good place to call this."

Do we call set_menubar() at the right time?

Also, the menu I get is the app menu. This a single menu named after the app. The docs say this:

"if both the application menu and the menubar are set, the application menu will be presented as if it were the first item of the menubar".
But I'm not setting 2 menus, only one. So this is strange.

**Reply by gtamas** [reply]
One more thing:

`println!("{:?}", app.list_action_descriptions());`

Calling this after set_menubar() prints this:

`["app.preferences", "gtkinternal.hide-others", "gtkinternal.hide", "app.quit", "win.about", "win.quit", "win.settings", "win.help"]`

The actions we just registered are not listed. Is that OK?

**Reply by AaronErhardt** [reply]
Well, you can try to keep the Relm4 code and just connect to the startup signal manually:

```rust
fn main() {
    let gtk_app = main_application();

    // Move menu creation here
    relm4::menu! { main_menu: ... }

    gtk_app.connect_startup(move |app| {
        app.set_menubar(Some(&main_menu));
    });
    
    // Keep the regular RelmApp initialization stuff here
}
```

**Reply by dominikwilkowski** [reply]
Tagging along here as I'm also trying to add a global menu bar for my app. The code above of the modified menu example posted by @AaronErhardt shows the menu items but they are indeed disabled on macos 14.1. A basic example of the global menu would be great. I'm not trying to do anything different here.

**Reply by gtamas** [reply]
I think it doesn't work. Menu items are always disabled no matter what. You also can't click to the menu unless you click outside the window first. That's a known GTK bug. But the menu items shouldn't be disabled, the example provided by GTK itself works (using only GTK code of course, no Relm4). This is MacOS 12.3.

**Reply by dominikwilkowski** [reply]
Should this be an issue then and move out of the discussions?

**Reply by gtamas** [reply]
OK I don't know how to do that. Should I close this and create an issue insted?
Who could test and reproduce this? Problem is Aaron doesn't have a mac as far as I know therefore he can't test this under OSX. I believe this is an OSX specific issue. The menu macros work fine under linux / unix.

**Reply by dominikwilkowski** [reply]
I'm happy to help in any way I can of course

**Reply by AaronErhardt** [reply]
> But the menu items shouldn't be disabled, the example provided by GTK itself works (using only GTK code of course, no Relm4)

If someone could create a gtk-rs example that works, that would help a lot. I think there might be some timing problems here (because Relm4 might register some stuff too late for MacOS), and to identify the problem in Relm4, we first need a working piece of code that does things correctly.

If you want to help with this, you could open an issue for that for further discussion.

**Reply by gtamas** [reply]
Will try to put together a working example and I'll post the link or code here.

**Reply by Moosems** [reply]
@gtamas We're you ever able to create the example?
